skfda/preprocessing/dim_reduction/_fpls.py

```diff
@@ -0,0 +1,829 @@
+from __future__ import annotations
+
+from typing import Any, Generic, Optional, Tuple, TypeVar, Union, cast
+
+import multimethod
+import numpy as np
+import scipy
+from sklearn.utils.validation import check_is_fitted
+
+from ..._utils._sklearn_adapter import BaseEstimator
+from ...misc.regularization import L2Regularization, compute_penalty_matrix
+from ...representation import FData, FDataGrid
+from ...representation.basis import Basis, FDataBasis, _GridBasis
+from ...typing._numpy import NDArrayFloat
+
+POWER_SOLVER_EPS = 1e-15
+INV_EPS = 1e-15
+
+
+def _power_solver(X: NDArrayFloat) -> NDArrayFloat:
```

Comment:
Yes. By around 5x. And, since this is one of the most performance-critical parts of the code, I think we should keep it as is. I have profiled the execution of the test `test_basis_vs_grid` in `test_fpls.py`. In particular, I have looked at the `_calculate_weights` function, which is the only one that calls `_power_solver`.

These are the profiling results using scipy's routine:
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    36                                           @profile                                                          
    37                                           def _calculate_weights(                                           
    38                                               X: NDArrayFloat,                                              
    39                                               Y: NDArrayFloat,                                              
    40                                               G_ww: NDArrayFloat,                                           
    41                                               G_xw: NDArrayFloat,                                           
    42                                               G_cc: NDArrayFloat,                                           
    43                                               G_yc: NDArrayFloat,                                           
    44                                               L_X_inv: NDArrayFloat,                                        
    45                                               L_Y_inv: NDArrayFloat,                                        
    46                                           ) -> Tuple[NDArrayFloat, NDArrayFloat]:                           
    47                                               """                                                           
    48                                               Calculate the weights for the PLS algorithm.                  
    49                                                                                                             
    50                                               Parameters as in _pls_nipals.                                 
    51                                               Returns:                                                      
    52                                                   - w: (n_features, 1)                                      
    53                                                       The X block weights.                                  
    54                                                   - c: (n_targets, 1)                                       
    55                                                       The Y block weights.                                  
    56                                               """                                                           
    57        10     195719.0  19571.9      1.7      X = X @ G_xw @ L_X_inv.T                                      
    58        10     198173.0  19817.3      1.8      Y = Y @ G_yc @ L_Y_inv.T                                      
    59        10     131111.0  13111.1      1.2      S = X.T @ Y                                                   
    60        10    9904302.0 990430.2     88.1      w = scipy.sparse.linalg.eigs(S @ S.T, k=1, which='LR')[1].real
    61        10         21.0      2.1      0.0      if len(w.shape) == 2:                                         
    62        10         16.0      1.6      0.0          w = w[:, 0]                                               
    63                                               # w = _power_solver(S @ S.T)                                  
    64                                                                                                             
    65                                               # Calculate the other weight                                  
    66        10     206366.0  20636.6      1.8      c = np.dot(Y.T, np.dot(X, w))                                 
    67                                                                                                             
    68                                               # Undo the transformation                                     
    69        10     121563.0  12156.3      1.1      w = L_X_inv.T @ w                                             
    70                                                                                                             
    71                                               # Normalize                                                   
    72        10     151654.0  15165.4      1.3      w /= np.sqrt(np.dot(w.T, G_ww @ w)) + INV_EPS                 
    73                                                                                                             
    74                                               # Undo the transformation                                     
    75        10     128151.0  12815.1      1.1      c = L_Y_inv.T @ c                                             
    76                                                                                                             
    77                                               # Normalize the other weight                                  
    78        10     208436.0  20843.6      1.9      c /= np.sqrt(np.dot(c.T, G_cc @ c)) + INV_EPS                 
    79                                                                                                             
    80        10         11.0      1.1      0.0      return w, c
```

These are the results using `_power_solver`
```
Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
    36                                           @profile                                                            
    37                                           def _calculate_weights(                                             
    38                                               X: NDArrayFloat,                                                
    39                                               Y: NDArrayFloat,                                                
    40                                               G_ww: NDArrayFloat,                                             
    41                                               G_xw: NDArrayFloat,                                             
    42                                               G_cc: NDArrayFloat,                                             
    43                                               G_yc: NDArrayFloat,                                             
    44                                               L_X_inv: NDArrayFloat,                                          
    45                                               L_Y_inv: NDArrayFloat,                                          
    46                                           ) -> Tuple[NDArrayFloat, NDArrayFloat]:                             
    47                                               """                                                             
    48                                               Calculate the weights for the PLS algorithm.                    
    49                                                                                                               
    50                                               Parameters as in _pls_nipals.                                   
    51                                               Returns:                                                        
    52                                                   - w: (n_features, 1)                                        
    53                                                       The X block weights.                                    
    54                                                   - c: (n_targets, 1)                                         
    55                                                       The Y block weights.                                    
    56                                               """                                                             
    57        10     251949.0  25194.9      9.7      X = X @ G_xw @ L_X_inv.T                                        
    58        10     230770.0  23077.0      8.9      Y = Y @ G_yc @ L_Y_inv.T                                        
    59        10     139497.0  13949.7      5.4      S = X.T @ Y                                                     
    60                                               # w = scipy.sparse.linalg.eigs(S @ S.T, k=1, which='LR')[1].real
    61                                               # if len(w.shape) == 2:                                         
    62                                               #     w = w[:, 0]                                               
    63        10    1928081.0 192808.1     74.0      w = _power_solver(S @ S.T)                                      
    64                                                                                                               
    65                                               # Calculate the other weight                                    
    66        10        866.0     86.6      0.0      c = np.dot(Y.T, np.dot(X, w))                                   
    67                                                                                                               
    68                                               # Undo the transformation                                       
    69        10      19856.0   1985.6      0.8      w = L_X_inv.T @ w                                               
    70                                                                                                               
    71                                               # Normalize                                                     
    72        10       8827.0    882.7      0.3      w /= np.sqrt(np.dot(w.T, G_ww @ w)) + INV_EPS                   
    73                                                                                                               
    74                                               # Undo the transformation                                       
    75        10      19631.0   1963.1      0.8      c = L_Y_inv.T @ c                                               
    76                                                                                                               
    77                                               # Normalize the other weight                                    
    78        10       7388.0    738.8      0.3      c /= np.sqrt(np.dot(c.T, G_cc @ c)) + INV_EPS                   
    79                                                                                                               
    80        10          5.0      0.5      0.0      return w, c
```